src/gg.regression.unity.bots/Editor/Scripts/ActionManager/RGActionAnalysis.cs

```diff
@@ -959,43 +933,116 @@ public override void VisitMethodDeclaration(MethodDeclarationSyntax node)
+        }
+        
         /// <summary>
-        /// Called while the analysis is running when actions are discovered in the code or resources.
-        /// The syntax node parameter is optional (may be null) if the action was not derived from code.
+        /// Adds the specified action associated with the given syntax node (optional).
+        /// If the ObjectType is not specified on the action (null), then it is automatically inferred from
+        /// the given syntax node.
         /// </summary>
         private void AddAction(RGGameAction action, SyntaxNode sourceNode)
         {
-            string path = string.Join("/", action.Paths[0]);
-            string currentPath = path;
-            int count = 1;
-            while (_rawActions.ContainsKey(currentPath))
+            if (action.ObjectType == null)
             {
-                ++count;
-                currentPath = path + " (" + count + ")";
+                var classDecl = sourceNode.Ancestors().OfType<ClassDeclarationSyntax>().FirstOrDefault();
+                if (classDecl == null) return;
+                Type objectType = FindType(_currentModel.GetDeclaredSymbol(classDecl));
+                if (typeof(MonoBehaviour).IsAssignableFrom(objectType))
+                {
+                    action.ObjectType = objectType;
+                }
             }
-            action.Paths[0] = currentPath.Split("/");
-            _rawActions.Add(currentPath, action);
-            if (sourceNode != null)
+            if (action.ObjectType != null)
             {
-                if (!_rawActionsByNode.TryGetValue(sourceNode, out var nodeActions))
+                string path = string.Join("/", action.Paths[0]);
+                if (_rawActions.TryAdd(path, action))
                 {
-                    nodeActions = new List<RGGameAction>();
-                    _rawActionsByNode.Add(sourceNode, nodeActions);
+                    if (sourceNode != null)
+                    {
+                        if (!_rawActionsByNode.TryGetValue(sourceNode, out var nodeActions))
+                        {
+                            nodeActions = new List<RGGameAction>();
+                            _rawActionsByNode.Add(sourceNode, nodeActions);
+                        }
+                        nodeActions.Add(action);
+                    }
+                    _changed = true;
+                }
+            }
+            else
+            {
+                // If an action was identified outside a MonoBehaviour (e.g. in a helper method), then
+                // it is added as an "unbound" action (i.e. it is unknown what component is listening for it).
+                // The action is associated with its containing method. Eventually, if a method invocation is found
+                // from a MonoBehaviour to the method that contained the actions, then copies of all the unbound actions
+                // will be associated with the containing MonoBehaviour.
```

Comment:
Appreciate the comments! It helps a lot for reviewing and later changes.

Reply:
(non-blocking) What happens if multiple MonoBehaviours use those actions? Will it allow mapping to multiple?

Reply:
Yes, if the same helper method is re-used in multiple places then copies of the action will be made for each of those locations.